### PR TITLE
Cursor position

### PR DESCRIPTION
**Pull request by perunt**
# Upstream PR Link

https://github.com/facebook/react-native/pull/36979
https://github.com/Expensify/App/issues/16078
## Summary
This PR extends the onSelectionChange method on textInput to return the x and y position of the input caret

## Changelog
In addition to the start and end indices of the selected text onSelectionChange returns also positionY and positionX  position of the input caret. The new object returned by the method looks like this:

```{
  selection: {
    start: number,
    end: number,
    positionY: number,
    positionX: number
  }
}
```

https://user-images.githubusercontent.com/48593211/233383277-2214564f-81b7-4552-9ed2-f1ad8d8b4d0b.mp4

positionX and positionY represent the coordinates inside the text input area, not including padding or margin. If you add a margin, you should add that extra space when using these coords to draw stuff on the text input.

[CATEGORY] [TYPE] - Message

## Test Plan

 added console.log statements to the onSelectionChange method to print the positionY and positionX values of the input caret. This allows for easy verification of the correct functioning of the new functionality added by this PR.

### CODE REVIEW

**Review by mountiny — APPROVED** [review]
The code looks good to me however, can we test this first? Could you build app against this commit/branch of our fork and test it there directly? Not sure how this testing was done before cc @roryabraham

**Comment by roryabraham** [comment]
@perunt It looks like this is missing an upstream PR.

For every change we make in our React Native fork, it's very important that we have:

1. An issue in the upstream React Native repo describing the problem we want to solve.
2. An upstream PR in the React Native repo that's essentially a duplicate of the PR we have in our fork.

Furthermore, we should try to get that upstream PR reviewed and merged before we merge changes in our fork, and only fall back on using our fork when that fails because review from Meta is taking too long.

**Review by roryabraham — CHANGES_REQUESTED** [review]
Also, new changes need to target the [ExpensifyRC1-0.72.0-alpha.0](https://github.com/Expensify/react-native/tree/ExpensifyRC1-0.72.0-alpha.0) branch

**Comment by perunt** [comment]
@mountiny, I can place an example in the `example` folder, allowing you to test it using RNTester.

**Comment by perunt** [comment]
@roryabraham Alright, I'll work on creating PR to RN. However, I'm not certain how quickly the RN team will respond. If this PR is held up for 2-6 months, would that also delay the implementation of this feature in our project for the same duration?

**Comment by mountiny** [comment]
> Alright, I'll work on creating PR to RN. However, I'm not certain how quickly the RN team will respond. If this PR is held up for 2-6 months, would that also delay the implementation of this feature in our project for the same duration?

No we would most likely not, but we want the issue up so we can drive towards updating this upstream too so we dont have to maintain the diff in our repo forever.

**Review by mczernek — APPROVED** [review]

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java [hunk 1]
```diff
@@ -1240,6 +1241,24 @@ public ReactSelectionWatcher(ReactEditText editText) {
...
+      if (mReactEditText.getLayout() == null) {
+            mReactEditText.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
+                @Override
+                public void onGlobalLayout() {
+                    mReactEditText.getViewTreeObserver().removeOnGlobalLayoutListener(this);
+                    onSelectionChanged(start, end);
+                }
+            });
+            return;
+        }
```
Comment (perunt) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java]: I observed that in version 0.71.6 of React Native, the mReactEditText variable can be null, so this is a hotfix to address that issue

**Review by mountiny — APPROVED** [review]
I think this is looking great now, thanks for updating the base version.

@roryabraham all yours

**Review by puneetlath — APPROVED** [review]

**Comment by puneetlath** [comment]
@roryabraham would you like to review this as well?

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java [hunk 2]
```diff
@@ -1240,6 +1241,24 @@ public ReactSelectionWatcher(ReactEditText editText) {
 
     @Override
     public void onSelectionChanged(int start, int end) {
+      // Calculate cursor position
+      Layout layout = mReactEditText.getLayout();
+      if (mReactEditText.getLayout() == null) {
+            mReactEditText.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
```
Comment (roryabraham) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java]: Minor thing, but I notice that the indentation of lines 1247 -> 1255 is off by two spaces

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java [hunk 3]
```diff
@@ -1240,6 +1241,24 @@ public ReactSelectionWatcher(ReactEditText editText) {
 
     @Override
     public void onSelectionChanged(int start, int end) {
+      // Calculate cursor position
+      Layout layout = mReactEditText.getLayout();
+      if (mReactEditText.getLayout() == null) {
```
Comment (roryabraham) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java]:
```suggestion

      // Wait for the TextInput to mount, if needed
      if (layout == null) {
```

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java [hunk 4]
```diff
@@ -1252,7 +1271,12 @@ public void onSelectionChanged(int start, int end) {
       if (mPreviousSelectionStart != realStart || mPreviousSelectionEnd != realEnd) {
         mEventDispatcher.dispatchEvent(
             new ReactTextInputSelectionEvent(
-                mSurfaceId, mReactEditText.getId(), realStart, realEnd));
+                mSurfaceId,
+                mReactEditText.getId(),
+                realStart,
```
Comment (roryabraham) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java]: Another minor thing, but there's an extraneous space at the end of line 1276 and 1277

**Review by roryabraham — CHANGES_REQUESTED** [review]
Just a few very minor points. Otherwise LGTM

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java (hunk 2 above)
Comment (perunt) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java]: 👍🏻

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java (hunk 4 above)
Comment (perunt) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java]: 👍🏻

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java (hunk 3 above)
Comment (perunt) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java]: I've made the update. Thank you!

**Review by roryabraham — APPROVED** [review]

**Comment by mountiny** [comment]
I think we are good to merge now.

**Pull request merged**